docs/grammar/SolidityLexer.g4

```diff
@@ -0,0 +1,335 @@
+Or: '||';
+And: '&&';
+BitOr: '|';
+BitXor: '^';
+BitAnd: '&';
+Shl: '<<';
+Sar: '>>';
+Shr: '>>>';
+Add: '+';
+Sub: '-';
+Mul: '*';
+Div: '/';
+Mod: '%';
+Exp: '**';
+
+Equal: '==';
+NotEqual: '!=';
+LessThan: '<';
+GreaterThan: '>';
+LessThanOrEqual: '<=';
+GreaterThanOrEqual: '>=';
+Not: '!';
+BitNot: '~';
+Inc: '++';
+Dec: '--';
+
+/**
+ * A single quoted string literal restricted to printable characters.
+ */
+StringLiteral: '"' DoubleQuotedStringCharacter* '"' | '\'' SingleQuotedStringCharacter* '\'';
+/**
+ * A single non-empty quoted string literal.
+ */
+NonEmptyStringLiteral: '"' DoubleQuotedStringCharacter+ '"' | '\'' SingleQuotedStringCharacter+ '\'';
+// Note that this will also be used for Yul string literals.
+//@doc:inline
+fragment DoubleQuotedStringCharacter: DoubleQuotedPrintable | EscapeSequence;
+// Note that this will also be used for Yul string literals.
+//@doc:inline
+fragment SingleQuotedStringCharacter: SingleQuotedPrintable | EscapeSequence;
+/**
+ * Any printable character expect single quote or back slash.
+ */
+fragment SingleQuotedPrintable: [\u0020-\u0026\u0028-\u005B\u005D-\u007E];
+/**
+ * Any printable character expect double quote or back slash.
+ */
+fragment DoubleQuotedPrintable: [\u0020-\u0021\u0023-\u005B\u005D-\u007E];
+/**
+  * Escape sequence.
+  * Apart from common single character escape sequences, line breaks can be escaped
+  * as well as four hex digit unicode escapes \\uXXXX and two digit hex escape sequences \\xXX are allowed.
+  */
+fragment EscapeSequence:
+    '\\' (
+        ['"\\bfnrtv\n\r]
+        | 'u' HexCharacter HexCharacter HexCharacter HexCharacter
+        | 'x' HexCharacter HexCharacter
+    );
+/**
+ * A single quoted string literal allowing arbitrary unicode characters.
+ */
+UnicodeStringLiteral:
+    'unicode"' DoubleQuotedUnicodeStringCharacter* '"'
+    | 'unicode\'' SingleQuotedUnicodeStringCharacter* '\'';
+//@doc:inline
+fragment DoubleQuotedUnicodeStringCharacter: ~["\r\n\\] | EscapeSequence;
+//@doc:inline
+fragment SingleQuotedUnicodeStringCharacter: ~['\r\n\\] | EscapeSequence;
+
+/**
+ * Hex strings need to consist of an even number of hex digits that may be grouped using underscores.
+ */
+HexString: 'hex' (('"' EvenHexDigits? '"') | ('\'' EvenHexDigits? '\''));
+/**
+ * Hex numbers consist of a prefix and an arbitrary number of hex digits that may be delimited by underscores.
+ */
+HexNumber: '0' 'x' HexDigits;
+//@doc:inline
+fragment HexDigits: HexCharacter ('_'? HexCharacter)*;
+//@doc:inline
+fragment EvenHexDigits: HexCharacter HexCharacter ('_'? HexCharacter HexCharacter)*;
+//@doc:inline
+fragment HexCharacter: [0-9A-Fa-f];
+
+/**
+ * A decimal number literal consists of decimal digits that may be delimited by underscores and
+ * an optional positive or negative exponent.
+ * If the digits contain a decimal point, the literal has fixed point type.
+ */
+DecimalNumber: (DecimalDigits | (DecimalDigits? '.' DecimalDigits)) ([eE] '-'? DecimalDigits)?;
```

Comment:
Is `(DecimalDigits '.' DecimalDigits?)` also allowed?

Reply:
Nope, parser error.